src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs

```diff
@@ -35,15 +36,40 @@ protected override object CreateTestClass()
         {
             var testClass = base.CreateTestClass();
 
-            if (testClass is ILoggedTest loggedTest)
+            (testClass as ILoggedTest).Initialize(
+                TestMethod,
+                TestMethodArguments,
+                _output ?? ConstructorArguments.SingleOrDefault(a => typeof(ITestOutputHelper).IsAssignableFrom(a.GetType())) as ITestOutputHelper);
+
+            return testClass;
+        }
+
+        protected override async Task<decimal> InvokeTestMethodAsync(object testClassInstance)
+        {
+            var loggedTestBase = testClassInstance as LoggedTestBase;
+            for (int i = 0; i < (loggedTestBase?.RetryLimit ?? 1); i++)
             {
-                loggedTest.Initialize(
-                    TestMethod,
-                    TestMethodArguments,
-                    _output ?? ConstructorArguments.SingleOrDefault(a => typeof(ITestOutputHelper).IsAssignableFrom(a.GetType())) as ITestOutputHelper);
+                if (i > 0)
+                {
+                    if (!loggedTestBase.RetryPredicate(Aggregator.ToException()))
+                    {
+                        return Timer.Total;
+                    }
+
+                    // This can only occur if the retry count has been overridden on the LoggedTestBase class
+                    loggedTestBase.Logger.LogWarning($"{TestMethod.Name} failed and retries are enabled, re-executing.");
+                }
+
+                Aggregator.Clear();
+                await base.InvokeTestMethodAsync(testClassInstance);
```

Comment:
At what level does this retry happen? It run-runs the method right? That might not actually work for some of the Kestrel tests that depend on state set up in the constructors. Have you been able to test it in Kestrel?

Reply:
Good catch I'll rerun the setup logic as well.